File: Mbs.Api/Extensions/ExceptionHandling/InnerError.cs
using System.Text.Json.Serialization;

namespace Mbs.Api.Extensions.ExceptionHandling
{
    /// <summary>
    /// The recursive inner error.
    /// </summary>
    public sealed class InnerError
    {
        /// <summary>
        /// Gets or sets the message of an inner error.
        /// </summary>
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
        public string Message { get; set; }

        /// <summary>
        /// Gets or sets inner errors of this inner error.
        /// </summary>
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
        public InnerError Details { get; set; }
    }
}

File: Mbs.Api.Host.Ng/ClientApp/src/app/shared/mbs/errors/inner-error.ts
/** A recursive inner error. */
export class InnerError {
  /** The message of an inner error. */
  message?: string | undefined;

  /** Inner errors of this inner error. */
  details?: InnerError | undefined;

  static appendMessage(text: string, innerError: InnerError): string {
    if (innerError.message) {
      text += innerError.message as string;
    }

    if (innerError.details) {
      text += ' (';
      text = InnerError.appendMessage(text, innerError.details as InnerError);
      text += ')';
    }

    return text;
  }
}
